src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,699 @@
+       return UCS_STATUS_PTR(status);
+    }
+
+    /* Start the request.
+     * If it is completed immediately, release the request and return the status.
+     * Otherwise, return the request.
+     */
+    status = ucp_request_send(req, 0);
+    if (req->flags & UCP_REQUEST_FLAG_COMPLETED) {
+        ucs_trace_req("releasing send request %p, returning status %s", req,
+                      ucs_status_string(status));
+        ucp_request_put(req);
+        return UCS_STATUS_PTR(status);
+    }
+
+    ucp_request_set_callback(req, send.cb, cb);
+    
+    return req + 1;
+}
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_send_nb, 
+                 (ep, id, payload, count, datatype, cb, flags),
+                 ucp_ep_h ep, uint16_t id, const void *payload, 
+                 size_t count, uintptr_t datatype, 
+                 ucp_send_callback_t cb, unsigned flags)
+{
+    ucs_status_t status;
+    ucs_status_ptr_t ret;
+    ucp_request_t *req;
+    size_t length;
+    
+    if (ucs_unlikely((flags != 0) && !(flags & UCP_AM_SEND_REPLY))) {
+        return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
+    }
+
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+    if (ucs_likely(!(flags & UCP_AM_SEND_REPLY)) && 
+        (ucs_likely(UCP_DT_IS_CONTIG(datatype)))) {
+        length = ucp_contig_dt_length(datatype, count);
+        
+        if (ucs_likely((ssize_t)length <= ucp_ep_config(ep)->am.max_short)) {
+            status = ucp_am_send_short(ep, id, payload, length);
+            if (ucs_likely(status != UCS_ERR_NO_RESOURCE)) {
+                UCP_EP_STAT_TAG_OP(ep, EAGER);
+                ret = UCS_STATUS_PTR(status);
+                goto out;
+            }
+        }
+    }
+
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(req == NULL)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
+    }
+
+    ucp_am_send_req_init(req, ep, payload, datatype, count, flags, id);
+    if (flags & UCP_AM_SEND_REPLY) {
+        status = ucp_ep_resolve_dest_ep_ptr(ep, ep->am_lane);
+        if (ucs_unlikely(status != UCS_OK)) {
+            ret = UCS_STATUS_PTR(status);
+            goto out;
+        }
+        ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                              ucp_ep_config(ep)->am_u.reply_proto);
+    } else {
+        ret = ucp_am_send_req(req, count, &ucp_ep_config(ep)->am, cb,
+                              ucp_ep_config(ep)->am_u.proto);
+    }
+
+out:
+    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(ep->worker);
+    return ret;
+}
+
+static ucs_status_t
+ucp_am_handler_common(ucp_worker_h worker, void *hdr_end,
+                      size_t hdr_size, size_t args_length,
+                      ucp_ep_h reply_ep, uint16_t am_id, 
+                      uint16_t desc_flag, unsigned am_flags) {
+    ucp_recv_desc_t *desc = NULL;
+    uint16_t recv_flags = 0;
+    ucs_status_t status;
+
+    if (ucs_unlikely((am_id >= worker->am_cb_array_len) ||
```

Comment:
looks like duplicate code with ucp_am_long_handler_common()

Reply:
Yes I will fix this.

Reply:
I thought this was referring to something else I had fixed previously. Yes this is common to both, should I make another common function, I think its not unreasonable to have it in both functions, but I'll change it if you want.

Reply:
let's try to fix in next PR to avoid code duplication (+fix indentation in ucp_am_long_handler_common)